docs/proposals/hpa/federated-hpa-v2.md

```diff
@@ -0,0 +1,809 @@
+}
+
+type TargetClusterAutoscalingPreference struct {
+	// TargetCluster describes the filter to select clusters.
+	// +required
+    TargetCluster           *policyv1alpha1.ClusterAffinity    `json:"targetCluster,omitempty"`
+
+    // AutoscalingMaxThreshold is the upper limit for the number of replicas for member clusters, it only works when the FederatedHPAAssignmentType is Aggregated/DynamicWeighted/Prioritized.
+    // +optional
+    // AutoscalingMaxThreshold int32                              `json:"autoscalingMaxThreshold,omitempty"`
+
+    // AutoscalingMinThreshold is the lower limit for the number of replicas for member clusters, it only works when the FederatedHPAAssignmentType is Aggregated/DynamicWeighted/Prioritized.
+    // +optional
+    // AutoscalingMinThreshold int32                              `json:"autoscalingMinThreshold,omitempty"`
+
+    // StaticWeight is the weight for the target cluster, it only works when the FederatedHPAAssignmentType is StaticWeighted.
+    // +optional
+    StaticWeight            *int32                             `json:"staticWeight,omitempty"`
+    
+    // Priority is the priority for the target cluster, it only works when the FederatedHPAAssignmentType is Prioritized.
+    // +optional
+    Priority                *int32                             `json:"priority,omitempty"`
+}
+
+type FederatedHPAStatus struct {
+	// AggregatedStatus contains the assign results and the status of HPA.
+    AggregatedStatus []ClusterHPAStatus `json:"aggregatedStatus,omitempty"`
+}
+
+type ClusterHPAStatus struct {
+    // LastScaleTime is the last time the HorizontalPodAutoscaler scaled the number of pods,
+    // used by the autoscaler to control how often the number of pods is changed.
+    // +optional
+    LastScaleTime *metav1.Time `json:"lastScaleTime,omitempty"`
+
+    // CurrentReplicas is current number of replicas of pods managed by this autoscaler,
+    // as last seen by the autoscaler.
+    CurrentReplicas int32 `json:"currentReplicas,omitempty"`
+
+    // DesiredReplicas is the desired number of replicas of pods managed by this autoscaler,
+    // as last calculated by the autoscaler.
+    DesignedReplicas int32 `json:"designedReplicas,omitempty"`
+
+    // CurrentMetrics is the last read state of the metrics used by this autoscaler.
+    // +optional
+    CurrentMetrics []autoscalingv2.MetricStatus `json:"currentMetrics,omitempty"`
+
+    // Conditions is the set of conditions required for this autoscaler to scale its target,
+    // and indicates whether or not those conditions are met.
+    Conditions []autoscalingv2.HorizontalPodAutoscalerCondition `json:"conditions,omitempty"`
+}
+
+```
+
+### Prerequisites
+1. Once the FederatedHPA is applied for the workload, Karmada scheduler should not work for this workload anymore.
+1. Once the FederatedHPA is applied for workload, the replicas changes in member clusters should be retained.
```

Comment:
According to this prerequisites, it seems that `rb.spec.clusters` does not take effect. But the policy case below shows that RB is updated by FederatedHPA controller. I'm confused about why we should handle this field `rb.spec.clusters`?

Reply:
To implement the initial assignment, we still need `rb` to generate `works`, which will be used to create/update the workloads in member clusters with specific replicas. So field `rb.spec.clusters` is still used.

Reply:
> o implement the initial assignment, we still need rb to generate works

Do you mean we should need the initial result to generate HPA objects in member clusters? After that, do we still need these results?

Reply:
> Do you mean we should need the initial result to generate HPA objects in member clusters?

For instance, the initial assignment result is:
```
#cluster 1
minReplicas: 3
maxReplicas: 5
#cluster 2
minReplicas: 1
maxReplicas: 2

# the current ResourceBinding is:
- name: cluster1
  replicas: 3
```
So initial assignment needs to update `ResourceBinding` to:
```
- name: cluster1
  replicas: 3
- name: cluster2
  replicas: 1
```
If not, the deployment/replicas in cluster2 will be 0, and couldn't be scaled up. `ResourceBinding` is just like a tool used to manage the initial replicas when FederatedHPA applied for the first time.

Reply:
Got it.